tests/0122-ssl_cb.c

```diff
@@ -0,0 +1,308 @@
+    }
+    if (!expect_ok) {
+        TEST_FAIL_LATER("passed rd_kafka_metadata, but expected to fail\n");
+    }
+    rd_kafka_metadata_destroy(md);
+    rd_kafka_destroy(rk);
+    return;
+}
+
+
+static void read_ssl_file_from_env(const char *envvar, char **dest, size_t *length) {
+    const char *path_val;
+    size_t szrt;
+    FILE *fh;
+
+    path_val = test_getenv(envvar, NULL);
+    if (!path_val) {
+        TEST_SKIP("Test requires %s env var\n", envvar);
+        return;
+    }
+    fh = fopen(path_val, "rb");
+    if (!fh) {
+        TEST_FAIL_LATER("File %s could not be opened: %m\n", path_val);
+        return;
+    }
+    fseek(fh, 0, SEEK_END);
+    *length = ftell(fh);
+    fseek(fh, 0, SEEK_SET);
+
+    *dest = malloc(*length);
+    // should probably handle short reads, but *shrug*
+    szrt = fread(*dest, 1, *length, fh);
+    if (!szrt) {
+        TEST_FAIL_LATER("File %s could not be read: %m\n", path_val);
+        fclose(fh);
+        return;
+    }
+    fclose(fh);
+    return;
+}
+
+int main_0122_ssl_cb(int argc, char **argv) {
+    struct test_context ctx;
+
+    memset(&ctx, 0, sizeof(struct test_context));
+
+    if (!test_check_builtin("ssl")) {
+        TEST_SKIP("Test requires SSL support\n");
+        goto cleanup;
+    }
+    if (!test_check_builtin("ssl_client_cert_callback")) {
+        TEST_SKIP("Test requires ssl_client_cert_callback support (i.e. built with openssl >= 1.0.2\n");
+        goto cleanup;
+    }
+
+
+    read_ssl_file_from_env(
+            "RDK_SSL_ca_pem", &ctx.ca_cert_pem, &ctx.ca_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_SSL_pub_pem", &ctx.direct_signed_client_cert_pem, &ctx.direct_signed_client_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_SSL_priv_pem", &ctx.direct_signed_client_key_pem, &ctx.direct_signed_client_key_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_UNTRUSTEDSSL_pub_pem", &ctx.self_signed_client_cert_pem, &ctx.self_signed_client_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_UNTRUSTEDSSL_priv_pem", &ctx.self_signed_client_key_pem, &ctx.self_signed_client_key_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_pub_pem", &ctx.intermediate_signed_client_cert_pem, &ctx.intermediate_signed_client_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_priv_pem", &ctx.intermediate_signed_client_key_pem, &ctx.intermediate_signed_client_key_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_intermediate_pub_pem", &ctx.intermediate_signed_intermediate_cert_pem, &ctx.intermediate_signed_intermediate_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_intermediate_priv_pem", &ctx.intermediate_signed_intermediate_key_pem, &ctx.intermediate_signed_intermediate_key_pem_length
+    );
+    if (test_curr->state == TEST_SKIPPED || test_curr->state == TEST_FAILED) {
```

Comment:
Test's shouldn't fiddle with this directly.
It is better to TEST_FAIL() immediately where the error is.

Reply:
So the reason I went to such effort to do it this way was because I wanted to make sure the test appropriately cleaned up after itself on fail. Memory and open file handles mostly. Seems like `TEST_FAIL` just calls `pthread_exit`, which seems rather violent :D

If the working assumption here is that we don't care about leaking in test cases, I can definitely go and just drop TEST_FAIL in all the right places.